### ISSUE TITLE
ICE: Complex control flow is forbidden in a const context

### ISSUE THREAD

**Issue by DutchGhost** [rust-lang/rust#69615]
I tried this code:

```rust
#![feature(const_trait_impl, const_fn)]

pub trait MyTrait {
    fn method(&self);
}

impl const MyTrait for std::convert::Infallible {
    #[inline(always)]
    fn method(&self) {
        match *self {}
    }
}
```

I expected to see this happen: Report an error in compilation which tells me that I need `#![feature(const_if_match)]`. If that feature is enabled, this ICE *DOESNT* happen.

Instead, this happened: This ICE

### Meta

`rustc --version --verbose`:
```
1.43.0-nightly (d3c79346a 2020-02-29)
```

<details><summary>Backtrace</summary>
<p>

```
warning: the feature `const_trait_impl` is incomplete and may cause the compiler to crash
 --> src/lib.rs:1:12
  |
1 | #![feature(const_trait_impl, const_fn)]
  |            ^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(incomplete_features)]` on by default

error: internal compiler error: complex control flow is forbidden in a const context
  --> src/lib.rs:9:5
   |
9  | /     fn method(&self) {
10 | |         match *self {}
11 | |     }
   | |_____^

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', src/librustc_errors/lib.rs:355:17
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1053
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1428
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:204
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:224
  10: rustc_driver::report_ice
  11: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
  12: std::panicking::begin_panic
  13: <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop
  14: core::ptr::drop_in_place
  15: <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop
  16: core::ptr::drop_in_place
  17: rustc_interface::interface::run_compiler_in_existing_thread_pool
  18: syntax::attr::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.43.0-nightly (d3c79346a 2020-02-29) running on x86_64-unknown-linux-gnu

note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `playground`.

```

</p>
</details>

**Comment by DutchGhost** [comment]
Edit:

This code ICE's as well, notice how this _does_ have the `const_if_match` feature enabled.
```Rust
#![feature(const_trait_impl, const_fn, const_if_match)]

pub trait MyTrait {
    fn method(&self);
}

impl const MyTrait for std::convert::Infallible {
    #[inline(always)]
    fn method(&self) {
        while let Some(_) = Option::<()>::None {}
    }
}
```

**Comment by ldm0** [comment]
> Edit:
> 
> This code ICE's as well, notice how this _does_ have the `const_if_match` feature enabled.
> 
> ```rust
> #![feature(const_trait_impl, const_fn, const_if_match)]
> 
> pub trait MyTrait {
>     fn method(&self);
> }
> 
> impl const MyTrait for std::convert::Infallible {
>     #[inline(always)]
>     fn method(&self) {
>         while let Some(_) = Option::<()>::None {}
>     }
> }
> ```

@DutchGhost This additional code not ICEs anymore.

**Comment by DutchGhost** [comment]
> 
> 
> > Edit:
> > This code ICE's as well, notice how this _does_ have the `const_if_match` feature enabled.
> > ```rust
> > #![feature(const_trait_impl, const_fn, const_if_match)]
> > 
> > pub trait MyTrait {
> >     fn method(&self);
> > }
> > 
> > impl const MyTrait for std::convert::Infallible {
> >     #[inline(always)]
> >     fn method(&self) {
> >         while let Some(_) = Option::<()>::None {}
> >     }
> > }
> > ```
> 
> @DutchGhost This additional code not ICEs anymore.

The issue has been fixed in https://github.com/rust-lang/rust/pull/69642
